v0/destinations/sendgrid/util.js

```diff
@@ -0,0 +1,152 @@
+const logger = require("../../../logger");
+const { CustomError } = require("../../util");
+
+const isValidBase64 = content => {
+  const re = /^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/;
+  return re.test(String(content));
+};
+
+const payloadValidator = payload => {
+  const updatedPayload = payload;
+  if (payload.personalizations.length < 1) {
+    throw new CustomError("personalization field cannot be empty", 400);
+  }
+  payload.personalizations.forEach((keys, index) => {
+    const personalizationsArr = [];
+    if (keys.to && (payload.subject || keys.subject)) {
+      keys.to.forEach(keyto => {
+        if (keyto.email) {
+          personalizationsArr.push(keyto);
+        }
+      });
+    } else {
+      logger.error(`item at index ${index} dropped. to field is mandatory`);
+    }
+    updatedPayload.personalizations[index].to = personalizationsArr;
+    if (keys.subject) {
+      updatedPayload.personalizations[index].subject = keys.subject;
+    }
+  });
+  if (payload.attachments) {
+    if (!payload.attachments.content || !payload.attachments.filename) {
+      updatedPayload.attachments = null;
+      logger.error("content and filename are required for attachments");
+    }
+    if (!isValidBase64(payload.attachments.content)) {
+      updatedPayload.attachments = null;
+      logger.error("content should be base64 encoded");
+    }
+  }
+  if (!payload.from.email) {
+    throw new CustomError("email is required inside from object", 400);
+  }
+  if (payload.categories) {
+    payload.categories.forEach((category, index) => {
+      if (typeof category !== "string") {
+        updatedPayload.categories[index] = String(category);
+      }
+    });
+    payload.categories.splice(10);
+  }
+  return updatedPayload;
+};
+
+const eventValidity = (Config, event) => {
```

Comment:
lets refactor this function. its still looping over the list even after flag=true is set